<a href="https://colab.research.google.com/github/ArmandoSep/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS_121_Statistics_Probability_and_Inference.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>

<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Lesson 1*

# Statistics, Probability and Inference

## Learning Objectives
* [Part 1](#p1): Normal Distribution Revisted
* [Part 2](#p2): Student's T Test
* [Part 3](#p3): Hypothesis Test & Doing it Live

## What is Descriptive Statistics?

<https://statistics.laerd.com/statistical-guides/descriptive-inferential-statistics.php>

In [0]:
import pandas as pd
df = pd.DataFrame({'a': [1,2,3,4,5], 'b': [2,4,6,8,10]})
df.head()

a   b
0  1   2
1  2   4
2  3   6
3  4   8
4  5  10

In [0]:
# How can we quickly look at some descriptive statistics of the above dataframe?
# All about describing a variable: its center, its spead, its distribution
df.describe()

a          b
count  5.000000   5.000000
mean   3.000000   6.000000
std    1.581139   3.162278
min    1.000000   2.000000
25%    2.000000   4.000000
50%    3.000000   6.000000
75%    4.000000   8.000000
max    5.000000  10.000000

## What is Inferential Statistics?

<https://slideplayer.com/slide/5130463/16/images/2/Statistical+Inference.jpg>

## Hypothesis Testing (T-Tests)

Drawing conclusions based on probability in the face of randomness



Is the sample reflective of our concept of the population?

In [0]:
# Coinflips example - What is the population?
import numpy as np



If we repeat the sampling why does our estimate of the mean move around so much? 

In [0]:
# Coinflips example - See the implications of randomness
# The probability of a coninflip is the population mean
conflips = np.random.binomial(n=1, p=.5, size=10)

conflips.mean()

0.7

What happens as I increase the sample size? The sample mean deviates less from the...?

In [0]:
# More Information -> Greater Confidence in the face of randomness
# The probability of a coninflip is the population mean
conflips = np.random.binomial(n=1, p=.5, size=1000)
# sample mean
conflips.mean()

0.516

# Standard Error

Quantifies how much our sample mean moves around. 

s = sample standard error

n = our sample size


As sample size increases, the SE goes down.

As sample size decreases, the SE goes up.

![Standard Error of the Sample  Mean](https://jungminded.weebly.com/uploads/1/3/5/8/13582285/7285766_orig.jpeg?160)

**Conclusion: As our sample size goes up, our estimates (of the mean in this case) get more precise, allowing us to infer things about the population with greater confidence.**



## One Sample T-test

What if I didn't have a fair coin?

Comparing information from a sample against our concept of the population.

![texto alternativo](https://lh3.googleusercontent.com/proxy/CXO8J4hOx3DqUt-vfRfmKN-B9IL0wNZlnbQ6lXnopiAAGQWtLIY-f4qD_2zCX_C3p_aaqi9-Qni2MqUD8mVbnnXWEy-AtAywlEp7bUZuMWIZwmvbFJFhdeHcUY5v-i3vcdg5cL9jdh6votwdHnFd44uE)
![One Sample T Statistic](https://lh3.googleusercontent.com/proxy/CXO8J4hOx3DqUt-vfRfmKN-B9IL0wNZlnbQ6lXnopiAAGQWtLIY-f4qD_2zCX_C3p_aaqi9-Qni2MqUD8mVbnnXWEy-AtAywlEp7bUZuMWIZwmvbFJFhdeHcUY5v-i3vcdg5cL9jdh6votwdHnFd44uE)

### Let's play a game
Get out a coin, and everytime that it is heads, I give Gwen a dollar.

Everytime that is tails, Gwen gives me a dollar.

## The T-test Process:

Working with an unfair coin here:

Probability of heads = .6

In [28]:
coinflips = np.random.binomial(n=1, p=.6, size=10)

coinflips

array([1, 1, 1, 1, 1, 0, 1, 1, 0, 1])

### Say that I came to the scenario not knowing if this was a fair coin or not.

How many times do i need to lose before I start get suspicious about this not being a gair coin?

Is there a way that I can mathetically test the fairness of this coin?

T-test: Compares a sample to a concept of a population, and tells me whether or not that sample is reflective of the population or not.

Concept of population: this is a fair coin. p=.5
Sample: ... We'll see what is says

## Establish the parts of a I=sample T-test

**Null Hypothesis**: This is a fair coin. My concept of the population is correct. $p(heads) == .5$

$H_0: \mu == \bar{x}$

My population mean {$\mu$} is equal to my sample mean {$\bar{x}$}

--

**Alternative Hypothesis**: {The opposite} My population mean is different than my sample mean

$H_a: \mu \neq \bar{x}$

My population mean {$\mu$} is not equal to my sample mean {$\bar{x}$}

--

**Confidence Level**: 95% Confidence Level: I am willingto be tricked by the randomness of the sampling process 5% of the time. We could draw the wrong conclusion 5% of the time due to the randomness of sampling.

This also means that I want to be 95% confidence that my null hypothesis is false before I reject it. 

I want to be less than 5% confident that my null hypothesis is true, before I will reject it. 

In [29]:
from scipy import stats

# Sample, Null Hypothesis Value (concept of the population, fair coin = .5)
# Is my sample reflective of my concept of the population
# if there's a low probability that my sample is representative of the population
# Then i will reject my null hypothesis that the coin is a fair coin.
stats.ttest_1samp(coinflips, .5)

Ttest_1sampResult(statistic=2.2500000000000004, pvalue=0.05100326070695074)

T-statistic: 2.25

**P-Value**: .05 -> Probability that the Null Hypothesis is true, given the sample that I have collected/observed

5.1%

Based on this p-value, should I reject or fail to reject the null hypothesis?

Critical Value = probablity threshold for reject vs fail to reject. (1-confidence leve) -> i.e. (1-.05)

P-Value >== .05 -> Fail to reject the null hypothesis

P-Value < .05 -> Reject the null hypotehesis

Conclusion: Based on p=value of .051 I fail to reject the null hypothesis that my sample mean of coinflips is equal to my population mean of coinflips. I cannot say that this isn't a fair coin.

In [35]:
# Increase sample size

coinflips = np.random.binomial(n=1, p=.6, size=50)
stats.ttest_1samp(coinflips, .5)

Ttest_1sampResult(statistic=2.041666666666667, pvalue=0.046587573163795186)

Conclusion: Based on p=value of .051 I reject the null hypothesis that my sample mean of coinflips is equal to my population mean of coinflips. There is less than a 5% chance that this is a fair coin given the sample that we observed. This is suggestive of the alternative hypothesis.



**The furder away the sample mean goes than our population goes. The lower the p-value will be.**

## Two Sample T-test

Comparing two groups or samples against each other rather than a single sample against a population. 



Ever thought about how long it takes to make a pancake? Have you ever compared the tooking time of a pancake on each eye of your stove? Is the cooking time different between the different eyes? Now, we can run an experiment and collect a sample of 1,000 pancakes on one eye and another 800 pancakes on the other eye. Assumed we used the same pan, batter, and technique on both eyes. Our average cooking times were 180 (5 std) and 178.5 (4.25 std) seconds repsectively. Now, we can tell those numbers are not identicial, but how confident are we that those numbers are practically the same? How do we know the slight difference isn't caused by some external randomness?

Yes, today's lesson will help you figure out how long to cook your pancakes (*theoretically*). Experimentation is up to you; otherwise, you have to accept my data as true. How are going to accomplish this? With probability, statistics, inference and maple syrup (optional). 

<img src="https://images.unsplash.com/photo-1541288097308-7b8e3f58c4c6?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=3300&q=80" width=400>



In [0]:
xbar1 = 180 # sample mean
s1 = 5 # sample standard deviation 1

sample1 = np.random.normal(xbar1, s1, 1000)
sample1

In [0]:
xbar2 = 178.5 # sample mean 2
s2 = 4.25 # sample standard deviation 2

sample2 = np.random.normal (xbar2, s2,800)
sample2

**Null Hypothesis**:

The cooking time of the two burners is the same.

$\bar(x)_1 == \bar(x)_2$ -> (Sample mean 1 is equal sample mean 2)

**Alternative Hypothesis:** 

The cooking time of the two burners is different

$\bar(x)_1 \neq \bar(x)_2$ -> (Sample mean 1 is not equal sample mean 2)


**Confidence level:** 95%

In [38]:
stats.ttest_ind(sample1, sample2)

Ttest_indResult(statistic=6.328265953929279, pvalue=3.1224446394725896e-10)

**Conclusion**: 

Based on T-statistic of 6.32 and a p-value of .00000000312, I **reject the null hypothesis** that the sample means are the same, and suggest the alternative, that the two burners coock at different rates. 

Reject when your P-Value < (1-Confidence level)

## Lets Practice more 1-sample and 2-sample T-tests

Using the dataset that you will be using on your assignment

In [40]:
# Let's do it!

!wget --no-check-certificate https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-04-08 21:58:01--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
  Issued certificate has expired.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.1s    

2020-04-08 21:58:01 (138 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [41]:
df = pd.read_csv('house-votes-84.data', names=['party','handicapped-infants','water-project',
                                                 'budget','physician-fee-freeze','el-salvador-aid',
                                                 'religious-groups','anti-satellite-ban',
                                                 'aid-t-contras','mx-aissile','immigration',
                                                 'synfuels','education','right-to-sue','crime','duty-free',
                                                 'south-africa'])

print (df.shape)
df.head()



(435, 17)


party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [42]:
df['party'].value_counts()

democrat      267
republican    168
Name: party, dtype: int64

In [44]:
# Label ecoding, and place ? with np.NaN in one fell-swoop.

df= df.replace({'y':1, 'n':0, '?': np.NaN})
df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [0]:
# split the dataset into republicans and democrats

rep = df[df['party']=='republican']
dem = df[df['party']=='democrat']

In [47]:
rep.head()

party  handicapped-infants  ...  duty-free  south-africa
0   republican                  0.0  ...        0.0           1.0
1   republican                  0.0  ...        0.0           NaN
7   republican                  0.0  ...        NaN           1.0
8   republican                  0.0  ...        0.0           1.0
10  republican                  0.0  ...        0.0           0.0

[5 rows x 17 columns]

In [48]:
dem.head()

party  handicapped-infants  water-project  ...  crime  duty-free  south-africa
2  democrat                  NaN            1.0  ...    1.0        0.0           0.0
3  democrat                  0.0            1.0  ...    0.0        0.0           1.0
4  democrat                  1.0            1.0  ...    1.0        1.0           1.0
5  democrat                  0.0            1.0  ...    1.0        1.0           1.0
6  democrat                  0.0            1.0  ...    1.0        1.0           1.0

[5 rows x 17 columns]

### 1 Sample t-test

When you're doing a 1-sample t-test, you get to pick what your null hypothesis value is. You get to pick what your conception of the population is. 

Republicans on the handicapped-infants bill 

**Null Hypothesis**: Republicans are split on this issue, 50% support, 50% do not support this bill. Republicans support, (the mean of yes-no votes on this bill for republicans is .5)

$H_0: \mu == \bar(x)$ ($\mu is .5)

I got to pick the .5

I could have said that I want to test that republican support is 30%, or 80%, or 0%. It's whatever I want to pick. 

By choosing you null hypothesis value, you get to frame the question that you're asking.

**Alternative:** $H_0: \mu \neq \bar(x)$ Republican support is not 50% (it's something other than 50%)

**Confidence level**: 95% (I'll compare my P-Value to the threshold of 0.05)

In [52]:
# I'll get nans for my t-stat and p-value if ther are NaNs in my sample
stats.ttest_1samp(rep['handicapped-infants'], .5)

Ttest_1sampResult(statistic=nan, pvalue=nan)

In [53]:
#therefore call nan_policy parameter to automatically drop the NaNs from the column
stats.ttest_1samp(rep['handicapped-infants'], .5, nan_policy='omit')

Ttest_1sampResult(statistic=-10.232833482397659, pvalue=2.572179359890009e-19)

Conclusion:

Reject the null hypothesis. It's likely that republican support is different than 50%

In [54]:
rep['handicapped-infants'].mean()

0.18787878787878787

# 2-sample t-test

Null hypothesis: the two samples are the same

Alternative: The two samples are different

(I do not get to choose the null hypothesis. It's always that they're the same in a 2-sample t-test.

Confidence level: 95%

In [55]:
stats.ttest_ind(rep['handicapped-infants'], dem['handicapped-infants'], nan_policy='omit')

Ttest_indResult(statistic=-9.205264294809222, pvalue=1.613440327937243e-18)

Conclusion: Based on p-value ~0%, I reject the null hypothesis that democrats and republicans support this bill at the same level - in same proportions. 

In [56]:
rep['handicapped-infants'].mean()

0.18787878787878787

In [57]:
dem['handicapped-infants'].mean()

0.6046511627906976

## Student's T Test

>Assuming data come from a Normal distribution, the t test provides a way to test whether the sample mean (that is the mean calculated from the data) is a good estimate of the population mean. 

The derivation of the t-distribution was first published in 1908 by William Gosset while working for the Guinness Brewery in Dublin. Due to proprietary issues, he had to publish under a pseudonym, and so he used the name Student.

The t-distribution is essentially a distribution of means of normaly distributed data. When we use a t-statistic, we are  checking that a mean fails within a certain $\alpha$ probability of the mean of means.

# Resources

- https://homepage.divms.uiowa.edu/~mbognar/applets/t.html
- https://rpsychologist.com/d3/tdist/
- https://gallery.shinyapps.io/tdist/
- https://en.wikipedia.org/wiki/Standard_deviation#Sample_standard_deviation_of_metabolic_rate_of_northern_fulmars
- https://www.khanacademy.org/math/ap-statistics/two-sample-inference/two-sample-t-test-means/v/two-sample-t-test-for-difference-of-means